### PR TITLE
More mmrest / copy/paste fixes

### PR DESCRIPTION
**Pull request by MarcSabatella**
(I was already doing further work on this branch when the original PR was merged; if I need to repeat this work on a new branch, that's fine, it won't be hard)

The original bug in https://musescore.org/en/node/74881 had to do with drawing the selection rectangle after a paste into an mmrest.  My initial fix addressed the symptom and successfully avoided the crash in that particular case but did not address the underlying cause and in testing I found other related issues, so I filed https://musescore.org/en/node/74946.

The first new commit here - the changes to Score::writeSegments() - fixes the most serious problems in 74946.  If a selection ends in an mmrest and we do a "copy", we were  actually copying the data all the way to the end of the score.  Apparently we manage to get away with this sometimes, but the issue report shows how to get corruption or a crash.  My fix detects these cases where we would have run past the end of the selection and accounts for them, hopefully without disturbing anything else.

The second commit addresses the less serious problem in 74946 (the selection after the paste going to the end of the score), which it turns out provides what might be considered a better fix for the original issue.  The relevant changes are the ones in Score::pasteStaff().  If mmrests are enabled, I force a relayout before trying to establish the new selection, and I use tick2segmentMM() to establish the start and end segments.  This should guarantee the start segment and end segment will not be within "hidden" (beneath mmrests) measures that have not been laid out yet, and would in theory mean my original fix is no longer required.

However, there may be other cases where the selection can get into that state, so better safe than sorry.  And while I was at it, I found other corner cases of crashes on pastes involving mmrests in this same place in the code that my original fix did not cover, so I improved that as well with the changes to ScoreView::paint().

So, basically, the first commit addresses the "copy" problem that led to the corruptions and crashes in 74946.  The changes to paste.cpp in the second commit addresses the root cause of the "paste" problem that led to the bad selection noted in 74946 as well as the original crash.  The changes to scoreview.cpp in that second commit are theoretically not needed with the changes to paste.cpp, but since there might be other ways to get into that situation, they seem good to keep.

Again, I am happy to repackage this as a separate branch, or squash the commits, or separate out the two changes in the second commit.

### CODE REVIEW

**Comment by lasconic** [comment]
Honnesly, this dolayout() here worries me... @wschweer  any insights?

**Comment by MarcSabatella** [comment]
I understand the concern.  I was worried about the possibility of some sort of infinite recursion, but of course there won't be another paste inside the doLayout().  Still, I share the general sense of uneasiness.

The reason I added it is so that the mmrests are correct when I update the selection - the result of the paste may well have changed them.  Without the doLayout(), I can't rely on tick2segmentMM() to return the correct results, and this seems to be really the key to fixing the basic problems here with updating the selection after the paste.  

FWIW, we also call doLayout() for what appear to be similar reasons here:

https://github.com/musescore/MuseScore/blob/a5cf31af659b4962baa11e62136f88b6be031e22/libmscore/line.cpp#L281

The actual work is done in Score::createMMRests(), we could try calling that instead of doLayout().  But I have no idea how safe that is outside of its usual context.  A quick test suggests it accomplishes the same thing, though.

Bottom line, though, it seems pretty broken right now - just about any paste into an mmrest produces an incorrect selection, and certain combinations produce crashes.  Calling either doLayout() or createMMRests() and then using tick2segmentMM() really does seem to fix everything to work smoothly.

**Comment by MarcSabatella** [comment]
I have updated the PR to use createMMRests() rather than doLayout().

BTW, not sure if it is coincidence or not, but this same code (the changes to paste.cpp) fixes https://musescore.org/en/node/75036, which was submitted shortly *after* this PR.

I can imagine there being lurking problems involving undo & redo.  But I have been doing lots of testing using the file attached to https://musescore.org/en/node/74946, and it does seem to work well.  If there are corner case issues remaining, they are probably not as bad as the current situation.

I guess a safer fix might be to have doLayout() itself update the selection after calling createMMRests, maybe checking to see if the end points of the selection are within "hidden" measures and replacing them by equivalent segments if so.  Seems that could come with its own problems, though.
